refactor tx building combinators

Grin node "transaction building combinators" are changing here -  https://github.com/mimblewimble/grin/pull/3057

We use some of these directly in grin-wallet when building basic transactions interactively.
These changes will require some (relatively minor) changes to grin-wallet to reflect these.

Example of building a tx via the new explicit tx kernel feature passing, replacing the existing `with_fee()` combinator - 

```
let tx = transaction(
    KernelFeatures::Plain { fee: 4 },
    vec![
        input(6, key_id1),
        output(2, key_id2),
    ],
    &keychain,
    &builder,
)
```

Related - https://github.com/mimblewimble/grin/pull/3057

Building grin-wallet against grin produces the following build errors (for reference) - 

```
error[E0425]: cannot find function `with_lock_height` in module `build`
   --> libwallet/src/internal/selection.rs:310:21
    |
310 |         parts.push(build::with_lock_height(lock_height));
    |                           ^^^^^^^^^^^^^^^^ not found in `build`

error[E0425]: cannot find function `with_fee` in module `build`
   --> libwallet/src/internal/selection.rs:447:20
    |
447 |     parts.push(build::with_fee(fee));
    |                       ^^^^^^^^ not found in `build`

error[E0061]: this function takes 4 parameters but 3 parameters were supplied
   --> libwallet/src/slate.rs:258:21
    |
258 |         let (tx, blind) = build::partial_transaction(elems, keychain, builder)?;
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 4 parameters

error: aborting due to 3 previous errors
```

`with_fee()` and `with_lock_height()` have been removed and replaced with explicitly passing the tx kernel (with corresponding kernel feature variant) into `build::partial_transaction()`

--- a/libwallet/src/internal/selection.rs
+++ b/libwallet/src/internal/selection.rs
@@ -15,7 +15,7 @@
 //! Selection of inputs for building transactions
 
 use crate::error::{Error, ErrorKind};
-use crate::grin_core::core::amount_to_hr_string;
+use crate::grin_core::core::{amount_to_hr_string, KernelFeatures, TxKernel};
 use crate::grin_core::libtx::{
 	build,
 	proof::{ProofBuild, ProofBuilder},
@@ -56,15 +56,23 @@ where
 		slate.amount,
 		slate.height,
 		minimum_confirmations,
-		slate.lock_height,
 		max_outputs,
 		change_outputs,
 		selection_strategy_is_use_all,
 		&parent_key_id,
 	)?;
-	let blinding = slate.add_transaction_elements(keychain, &ProofBuilder::new(keychain), elems)?;
 
-	slate.fee = fee;
+	// Update the fee on both the slate *and* the underlying tx.
+	// Need to think through how we would support lock_height here.
+	{
+		slate.fee = fee;
+		slate.tx = slate
+			.tx
+			.clone()
+			.replace_kernel(TxKernel::with_features(KernelFeatures::Plain { fee }));
+	}
+
+	let blinding = slate.add_transaction_elements(keychain, &ProofBuilder::new(keychain), elems)?;
 
 	// Create our own private context
 	let mut context = Context::new(
@@ -270,7 +278,6 @@ pub fn select_send_tx<'a, T: ?Sized, C, K, B>(
 	amount: u64,
 	current_height: u64,
 	minimum_confirmations: u64,
-	lock_height: u64,
 	max_outputs: usize,
 	change_outputs: usize,
 	selection_strategy_is_use_all: bool,
@@ -302,13 +309,8 @@ where
 	)?;
 
 	// build transaction skeleton with inputs and change
-	let (mut parts, change_amounts_derivations) =
+	let (parts, change_amounts_derivations) =
 		inputs_and_change(&coins, wallet, keychain_mask, amount, fee, change_outputs)?;
-
-	// Build a "Plain" kernel unless lock_height>0 explicitly specified.
-	if lock_height > 0 {
-		parts.push(build::with_lock_height(lock_height));
-	}
 
 	Ok((parts, coins, change_amounts_derivations, fee))
 }
@@ -443,8 +445,6 @@ where
 
 	// calculate the total across all inputs, and how much is left
 	let total: u64 = coins.iter().map(|c| c.value).sum();
-
-	parts.push(build::with_fee(fee));
 
 	// if we are spending 10,000 coins to send 1,000 then our change will be 9,000
 	// if the fee is 80 then the recipient will receive 1000 and our change will be

--- a/libwallet/src/internal/selection.rs
+++ b/libwallet/src/internal/selection.rs
@@ -15,7 +15,7 @@
 //! Selection of inputs for building transactions
 
 use crate::error::{Error, ErrorKind};
-use crate::grin_core::core::{amount_to_hr_string, KernelFeatures, TxKernel};
+use crate::grin_core::core::amount_to_hr_string;
 use crate::grin_core::libtx::{
 	build,
 	proof::{ProofBuild, ProofBuilder},
@@ -62,15 +62,8 @@ where
 		&parent_key_id,
 	)?;
 
-	// Update the fee on both the slate *and* the underlying tx.
-	// Need to think through how we would support lock_height here.
-	{
-		slate.fee = fee;
-		slate.tx = slate
-			.tx
-			.clone()
-			.replace_kernel(TxKernel::with_features(KernelFeatures::Plain { fee }));
-	}
+	// Update the fee on the slate so we account for this when building the tx.
+	slate.fee = fee;
 
 	let blinding = slate.add_transaction_elements(keychain, &ProofBuilder::new(keychain), elems)?;
 

--- a/libwallet/src/internal/tx.rs
+++ b/libwallet/src/internal/tx.rs
@@ -365,6 +365,7 @@ where
 
 #[cfg(test)]
 mod test {
+	use crate::grin_core::core::KernelFeatures;
 	use crate::grin_core::libtx::{build, ProofBuilder};
 	use crate::grin_keychain::{ExtKeychain, ExtKeychainPath, Keychain};
 
@@ -377,12 +378,14 @@ mod test {
 		let key_id1 = ExtKeychainPath::new(1, 1, 0, 0, 0).to_identifier();
 
 		let tx1 = build::transaction(
+			KernelFeatures::Plain { fee: 0 },
 			vec![build::output(105, key_id1.clone())],
 			&keychain,
 			&builder,
 		)
 		.unwrap();
 		let tx2 = build::transaction(
+			KernelFeatures::Plain { fee: 0 },
 			vec![build::input(105, key_id1.clone())],
 			&keychain,
 			&builder,

--- a/libwallet/src/slate.rs
+++ b/libwallet/src/slate.rs
@@ -245,17 +245,26 @@ impl Slate {
 		&mut self,
 		keychain: &K,
 		builder: &B,
-		mut elems: Vec<Box<build::Append<K, B>>>,
+		elems: Vec<Box<build::Append<K, B>>>,
 	) -> Result<BlindingFactor, Error>
 	where
 		K: Keychain,
 		B: ProofBuild,
 	{
-		elems.insert(0, build::initial_tx(self.tx.clone()));
-		let (tx, blind) =
-			build::partial_transaction(self.kernel_features(), elems, keychain, builder)?;
+		self.update_kernel();
+		let (tx, blind) = build::partial_transaction(self.tx.clone(), elems, keychain, builder)?;
 		self.tx = tx;
 		Ok(blind)
+	}
+
+	// Update the tx kernel based on kernel features derived from the current slate.
+	// The fee may change as we build a transaction and we need to
+	// update the tx kernel to reflect this during the tx building process.
+	pub fn update_kernel(&mut self) {
+		self.tx = self
+			.tx
+			.clone()
+			.replace_kernel(TxKernel::with_features(self.kernel_features()));
 	}
 
 	/// Completes callers part of round 1, adding public key info

--- a/libwallet/src/slate.rs
+++ b/libwallet/src/slate.rs
@@ -251,11 +251,9 @@ impl Slate {
 		K: Keychain,
 		B: ProofBuild,
 	{
-		// Append to the exiting transaction
-		if self.tx.kernels().len() != 0 {
-			elems.insert(0, build::initial_tx(self.tx.clone()));
-		}
-		let (tx, blind) = build::partial_transaction(elems, keychain, builder)?;
+		elems.insert(0, build::initial_tx(self.tx.clone()));
+		let (tx, blind) =
+			build::partial_transaction(self.kernel_features(), elems, keychain, builder)?;
 		self.tx = tx;
 		Ok(blind)
 	}
@@ -290,17 +288,22 @@ impl Slate {
 		Ok(())
 	}
 
-	// This is the msg that we will sign as part of the tx kernel.
-	// If lock_height is 0 then build a plain kernel, otherwise build a height locked kernel.
-	fn msg_to_sign(&self) -> Result<secp::Message, Error> {
-		let features = match self.lock_height {
+	// Construct the appropriate kernel features based on our fee and lock_height.
+	// If lock_height is 0 then its a plain kernel, otherwise its a height locked kernel.
+	fn kernel_features(&self) -> KernelFeatures {
+		match self.lock_height {
 			0 => KernelFeatures::Plain { fee: self.fee },
 			_ => KernelFeatures::HeightLocked {
 				fee: self.fee,
 				lock_height: self.lock_height,
 			},
-		};
-		let msg = features.kernel_sig_msg()?;
+		}
+	}
+
+	// This is the msg that we will sign as part of the tx kernel.
+	// If lock_height is 0 then build a plain kernel, otherwise build a height locked kernel.
+	fn msg_to_sign(&self) -> Result<secp::Message, Error> {
+		let msg = self.kernel_features().kernel_sig_msg()?;
 		Ok(msg)
 	}
 
@@ -501,6 +504,7 @@ impl Slate {
 			self.tx.kernels().len(),
 			None,
 		);
+
 		if fee > self.tx.fee() {
 			return Err(ErrorKind::Fee(
 				format!("Fee Dispute Error: {}, {}", self.tx.fee(), fee,).to_string(),

--- a/libwallet/src/slate.rs
+++ b/libwallet/src/slate.rs
@@ -257,9 +257,9 @@ impl Slate {
 		Ok(blind)
 	}
 
-	// Update the tx kernel based on kernel features derived from the current slate.
-	// The fee may change as we build a transaction and we need to
-	// update the tx kernel to reflect this during the tx building process.
+	/// Update the tx kernel based on kernel features derived from the current slate.
+	/// The fee may change as we build a transaction and we need to
+	/// update the tx kernel to reflect this during the tx building process.
 	pub fn update_kernel(&mut self) {
 		self.tx = self
 			.tx

